CosmosDB polling/set fixes

Opening this issue to ensure it's tracked for the release. The following CosmosDB tests need fixing before we release 1.9.0:

```
------- Stdout: -------
=== RUN   TestAccAzureRMCosmosDBAccount_failover_boundedStaleness
--- FAIL: TestAccAzureRMCosmosDBAccount_failover_boundedStaleness (930.25s)
	testing.go:573: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.
		
		Error: Error applying: 1 error(s) occurred:
		
		* azurerm_resource_group.test (destroy): 1 error(s) occurred:
		
		* azurerm_resource_group.test: Error deleting Resource Group "acctestRG-3021637299523799643": pollingTrackerBase#updatePollingState: the response from the async operation has an invalid status code: StatusCode=0
		
		State: azurerm_resource_group.test:
		  ID = /subscriptions/*******/resourceGroups/acctestRG-3021637299523799643
		  provider = provider.azurerm
		  location = westeurope
		  name = acctestRG-3021637299523799643
		  tags.% = 0
FAIL
```

```
------- Stdout: -------
=== RUN   TestAccDataSourceAzureRMCosmosDBAccount_complete
--- FAIL: TestAccDataSourceAzureRMCosmosDBAccount_complete (607.20s)
	testing.go:513: Step 0 error: Error applying: 1 error(s) occurred:
		
		* data.azurerm_cosmosdb_account.test: data.azurerm_cosmosdb_account.test: Error setting `capabilities`: capabilities: '': source data must be an array or slice, got struct
	testing.go:573: Error destroying resource! WARNING: Dangling resources
		may exist. The full state and error is shown below.
		
		Error: Error refreshing: 1 error(s) occurred:
		
		* data.azurerm_cosmosdb_account.test: 1 error(s) occurred:
		
		* data.azurerm_cosmosdb_account.test: data.azurerm_cosmosdb_account.test: Error setting `capabilities`: capabilities: '': source data must be an array or slice, got struct
		
		State: <nil>
FAIL
```

--- a/azurerm/data_source_cosmos_db_account.go
+++ b/azurerm/data_source_cosmos_db_account.go
@@ -4,6 +4,8 @@ import (
 	"fmt"
 	"log"
 
+	"github.com/Azure/azure-sdk-for-go/services/cosmos-db/mgmt/2015-04-08/documentdb"
+
 	"github.com/hashicorp/terraform/helper/schema"
 	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
 )
@@ -95,9 +97,14 @@ func dataSourceArmCosmosDBAccount() *schema.Resource {
 
 			"capabilities": {
 				Type:     schema.TypeList,
-				Optional: true,
-				Elem: &schema.Schema{
-					Type: schema.TypeString,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
 				},
 			},
 
@@ -198,7 +205,7 @@ func dataSourceArmCosmosDBAccountRead(d *schema.ResourceData, meta interface{}) 
 			return fmt.Errorf("Error setting `geo_location`: %+v", err)
 		}
 
-		if err := d.Set("capabilities", flattenAzureRmCosmosDBAccountCapabilities(resp.Capabilities)); err != nil {
+		if err := d.Set("capabilities", flattenAzureRmCosmosDBAccountCapabilitiesAsList(resp.Capabilities)); err != nil {
 			return fmt.Errorf("Error setting `capabilities`: %+v", err)
 		}
 
@@ -236,4 +243,19 @@ func dataSourceArmCosmosDBAccountRead(d *schema.ResourceData, meta interface{}) 
 	}
 
 	return nil
+}
+
+func flattenAzureRmCosmosDBAccountCapabilitiesAsList(capabilities *[]documentdb.Capability) *[]map[string]interface{} {
+	slice := make([]map[string]interface{}, 0)
+
+	for _, c := range *capabilities {
+		if v := c.Name; v != nil {
+			e := map[string]interface{}{
+				"name": *v,
+			}
+			slice = append(slice, e)
+		}
+	}
+
+	return &slice
 }

--- a/azurerm/data_source_cosmos_db_account_test.go
+++ b/azurerm/data_source_cosmos_db_account_test.go
@@ -66,7 +66,7 @@ func TestAccDataSourceAzureRMCosmosDBAccount_complete(t *testing.T) {
 				Config: testAccDataSourceAzureRMCosmosDBAccount_complete(ri, testLocation(), testAltLocation()),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					checkAccAzureRMCosmosDBAccount_basic(dataSourceName, testLocation(), string(documentdb.BoundedStaleness), 2),
-					resource.TestCheckResourceAttr(dataSourceName, "ip_range_filter", "104.42.195.92,40.76.54.131,52.176.6.30,52.169.50.45,52.187.184.26"),
+					resource.TestCheckResourceAttr(dataSourceName, "ip_range_filter", "104.42.195.92,40.76.54.131,52.176.6.30,52.169.50.45,52.187.184.26,10.20.0.0/16"),
 					resource.TestCheckResourceAttr(dataSourceName, "enable_automatic_failover", "true"),
 					resource.TestCheckResourceAttr(dataSourceName, "geo_location.0.location", testLocation()),
 					resource.TestCheckResourceAttr(dataSourceName, "geo_location.1.location", testAltLocation()),

--- a/azurerm/resource_arm_cosmos_db_account.go
+++ b/azurerm/resource_arm_cosmos_db_account.go
@@ -573,10 +573,10 @@ func resourceArmCosmosDBAccountDelete(d *schema.ResourceData, meta interface{}) 
 		return err
 	}
 
-	resGroup := id.ResourceGroup
+	resourceGroup := id.ResourceGroup
 	name := id.Path["databaseAccounts"]
 
-	future, err := client.Delete(ctx, resGroup, name)
+	future, err := client.Delete(ctx, resourceGroup, name)
 	if err != nil {
 		if response.WasNotFound(future.Response()) {
 			return nil
@@ -584,13 +584,37 @@ func resourceArmCosmosDBAccountDelete(d *schema.ResourceData, meta interface{}) 
 		return fmt.Errorf("Error issuing AzureRM delete request for CosmosDB Account '%s': %+v", name, err)
 	}
 
-	err = future.WaitForCompletion(ctx, client.Client)
+	err = future.WaitForCompletionRef(ctx, client.Client)
 	if err != nil {
-		if response.WasNotFound(future.Response()) {
-			return nil
+		if !response.WasNotFound(future.Response()) {
+			return fmt.Errorf("Error issuing AzureRM delete request for CosmosDB Account '%s': %+v", name, err)
 		}
-		return fmt.Errorf("Error issuing AzureRM delete request for CosmosDB Account '%s': %+v", name, err)
 	}
+
+	//the SDK now will return a `WasNotFound` response even when still deleting
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"Deleting"},
+		Target:     []string{"NotFound"},
+		Timeout:    60 * time.Minute,
+		MinTimeout: 30 * time.Second,
+		Refresh: func() (interface{}, string, error) {
+
+			resp, err := client.Get(ctx, resourceGroup, name)
+			if err != nil {
+
+				if utils.ResponseWasNotFound(resp.Response) {
+					return resp, "NotFound", nil
+				}
+				return nil, "", fmt.Errorf("Error reading CosmosDB Account %q after delete (Resource Group %q): %+v", name, resourceGroup, err)
+			}
+
+			return resp, "Deleting", nil
+		},
+	}
+	if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf("Waiting forCosmosDB Account %q to delete (Resource Group %q): %+v", name, resourceGroup, err)
+	}
+
 	return nil
 }
 

--- a/azurerm/resource_arm_cosmos_db_account.go
+++ b/azurerm/resource_arm_cosmos_db_account.go
@@ -584,13 +584,6 @@ func resourceArmCosmosDBAccountDelete(d *schema.ResourceData, meta interface{}) 
 		return fmt.Errorf("Error issuing AzureRM delete request for CosmosDB Account '%s': %+v", name, err)
 	}
 
-	err = future.WaitForCompletionRef(ctx, client.Client)
-	if err != nil {
-		if !response.WasNotFound(future.Response()) {
-			return fmt.Errorf("Error issuing AzureRM delete request for CosmosDB Account '%s': %+v", name, err)
-		}
-	}
-
 	//the SDK now will return a `WasNotFound` response even when still deleting
 	stateConf := &resource.StateChangeConf{
 		Pending:    []string{"Deleting"},

